util/topsql/reporter/reporter.go

```diff
@@ -0,0 +1,411 @@
+	PlanBinaryDecoder planBinaryDecodeFunc
+	// MaxStatementsNum is the capacity of the TopSQL map, above which evition should happen
+	MaxStatementsNum int
+	// CollectInterval is the interval of sending TopSQL records to the agent
+	CollectInterval time.Duration
+	// CollectTimeout is the timeout of a single agent gRPC call
+	CollectTimeout time.Duration
+	// AgentGRPCAddress is the gRPC address of the agent server
+	AgentGRPCAddress string
+	// InstanceID is the ID of this tidb server
+	InstanceID string
+}
+
+func encodeCacheKey(sqlDigest, planDigest []byte) []byte {
+	var buffer bytes.Buffer
+	buffer.Write(sqlDigest)
+	buffer.Write(planDigest)
+	return buffer.Bytes()
+}
+
+const (
+	grpcInitialWindowSize     = 1 << 30
+	grpcInitialConnWindowSize = 1 << 30
+)
+
+func newAgentClient(addr string) (*grpc.ClientConn, tipb.TopSQLAgentClient, error) {
+	dialCtx, cancel := context.WithTimeout(context.TODO(), time.Second)
+	defer cancel()
+	conn, err := grpc.DialContext(
+		dialCtx,
+		addr,
+		grpc.WithInsecure(),
+		grpc.WithInitialWindowSize(grpcInitialWindowSize),
+		grpc.WithInitialConnWindowSize(grpcInitialConnWindowSize),
+		grpc.WithDefaultCallOptions(
+			grpc.MaxCallRecvMsgSize(math.MaxInt64),
+		),
+		grpc.WithConnectParams(grpc.ConnectParams{
+			Backoff: backoff.Config{
+				BaseDelay:  100 * time.Millisecond, // Default was 1s.
+				Multiplier: 1.6,                    // Default
+				Jitter:     0.2,                    // Default
+				MaxDelay:   3 * time.Second,        // Default was 120s.
+			},
+		}),
+	)
+	if err != nil {
+		return nil, nil, err
+	}
+	client := tipb.NewTopSQLAgentClient(conn)
+	return conn, client, nil
+}
+
+// NewTopSQLReporter creates a new TopSQL struct
+//
+// planBinaryDecoder is a decoding function which will be called asynchronously to decode the plan binary to string
+// MaxStatementsNum is the maximum SQL and plan number, which will restrict the memory usage of the internal LFU cache
+func NewTopSQLReporter(config *TopSQLReporterConfig) *TopSQLReporterImpl {
+	collectTimeout := config.CollectTimeout
+	if collectTimeout > config.CollectInterval || collectTimeout <= 0 {
+		collectTimeout = config.CollectInterval / 2
+	}
+	tsc := &TopSQLReporterImpl{
+		planBinaryDecoder:  config.PlanBinaryDecoder,
+		MaxStatementsNum:   config.MaxStatementsNum,
+		collectInterval:    config.CollectInterval,
+		collectTimeout:     collectTimeout,
+		topSQLMap:          make(map[string]*topSQLDataPoints),
+		normalizedSQLMap:   make(map[string]string),
+		normalizedPlanMap:  make(map[string]string),
+		collectCPUTimeChan: make(chan *topSQLCPUTimeInput, 1),
+		planRegisterChan:   make(chan *planRegisterJob, 10),
+		agentGRPCAddress:   config.AgentGRPCAddress,
+		instanceID:         config.InstanceID,
+		quit:               make(chan struct{}),
+	}
+
+	go tsc.collectWorker()
+
+	go tsc.registerPlanWorker()
+
+	go tsc.sendToAgentWorker()
+
+	return tsc
+}
+
+// Collect will drop the records when the collect channel is full
```

Comment:
TODO: You can add a test to verify this behaviour later.

Reply:
LGTM